src/algorithms.rs

```diff
@@ -52,24 +56,60 @@ pub enum Error {
 
     /// When a partition improving algorithm is given more than 2 parts.
     BiPartitioningOnly,
+
+    /// Conversion between types failed.
+    Conversion {
+        src_type: &'static str,
+        dst_type: &'static str,
+    },
 }
 
 impl fmt::Display for Error {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         match self {
-            Error::NotFound => write!(f, "no partition found"),
-            Error::InputLenMismatch { expected, actual } => write!(
+            Self::Alloc => write!(f, "out of memory"),
+            Self::NotFound => write!(f, "no partition found"),
+            Self::InputLenMismatch { expected, actual } => write!(
                 f,
                 "input sets don't have the same length (expected {expected} items, got {actual})",
             ),
-            Error::NegativeValues => write!(f, "input contains negative values"),
-            Error::BiPartitioningOnly => write!(f, "expected no more than two parts"),
+            Self::NegativeValues => write!(f, "input contains negative values"),
+            Self::BiPartitioningOnly => write!(f, "expected no more than two parts"),
+            Self::Conversion { src_type, dst_type } => {
+                write!(f, "failed conversion from {src_type} to {dst_type}")
+            }
         }
     }
 }
 
 impl std::error::Error for Error {}
 
+impl From<TryReserveError> for Error {
+    fn from(_: TryReserveError) -> Self {
+        Self::Alloc
+    }
+}
+
+fn try_from_f64<T>(f: f64) -> Result<T, Error>
+where
+    T: num::FromPrimitive,
+{
+    T::from_f64(f).ok_or_else(|| Error::Conversion {
+        src_type: std::any::type_name::<f64>(),
+        dst_type: std::any::type_name::<T>(),
+    })
+}
+
+fn try_to_f64<T>(t: &T) -> Result<f64, Error>
+where
+    T: num::ToPrimitive,
+{
+    T::to_f64(t).ok_or_else(|| Error::Conversion {
+        src_type: std::any::type_name::<T>(),
+        dst_type: std::any::type_name::<f64>(),
+    })
+}
```

Comment:
Instead of `num::{To,From}Primitive`, there is `num_traits::AsPrimitive`, which uses `as` and does not return an Option.

Downside is it cannot be as easily used with the `AlgWeight` trait aliases:

```rust
trait SomeAlgWeight
where
    Self: AsPrimitive<f64>,
    f64: AsPrimitive<Self>,
{}

fn call_alg<T>()
where
    T: SomeAlgWeight,
    f64: AsPrimtive<T>, // this need to appear at all call sites, not ideal
{}
```